File: CS254/assign2/Java/Stack.java
/*
 * Title:	Stack.java
 * Author:	Dilpreet S. Chana
 * A General Stack ADT using ArrayList
**/

import java.util.*;

public class Stack<T>{
	private ArrayList<T> nodes;
	private T top;
	private int size;

	// Constructor
	public Stack(){
		nodes = new ArrayList<T>();
		top = null;
		size = 0;
	}

	public T getTop(){
		return top;
	}

	public void push(T node){
		nodes.add(node);
		top = nodes.get(size);
		size ++;
	}
	public T pop(){
		T tmp = top;

		if(size > 0){
			nodes.remove(size - 1);
			size --;

			if(size > 0) top = nodes.get(size - 1);
			else top = null;
		}
		return tmp;
	}

	@Override
	public String toString(){
		String val = "";

		for(T cur : nodes){
			val += cur.toString() + " ";
		}
		return val;
	}
}

File: CS254/assign2/Stack.cpp
#include <Stack.h>
#include <vector>
#include <sstream>
#include <string>

using namespace std;

Stack::Stack() {
	size = 0;
}

void Stack::push(T n) {
	elements.push_back(n);
	size++;
}

T Stack::pop() {
	if (size > 0) {
		T tmp = elements[size - 1];
		elements.erase(elements.end());
		size--;

		return tmp;
	}

	return NULL;
}

T Stack::getTop() {
	if (size > 0) {
		return elements[size - 1];
	}

	return NULL;
}

char* Stack::to_string() {
	stringstream ss;

	for (int i = 0; i < size; i++) {
		ss << elements[i] << ' ';
	}

	return ss.str().c_str();
}
